geopandas/tools/sjoin.py

```diff
@@ -117,91 +115,29 @@ def sjoin(
         right_df.index = right_df.index.rename(index_right)
     right_df = right_df.reset_index()
 
-    if op == "within":
-        # within implemented as the inverse of contains; swap names
-        left_df, right_df = right_df, left_df
-        tree_idx_right = not tree_idx_right
+    # for historical reasons, this logic is flipped in sjoin vs. pygeos query_bulk
+    if op == "contains":
+        op = "within"
+    elif op == "within":
+        op = "contains"
```

Comment:
This part confuses me a bit. It would be great if someone could confirm if I am right or wrong.

Reply:
As far as I understand, this would need to depend on whether the spatial index of the left or right frame is being used

(in general, this flexibility to either use left or right frame's sindex seems a complication here, certainly if we want to enable more predicate ops)

Reply:
Right, which is what confuses me as far as what the expected result is. The above logic works, but it's quite strange to have to flip the predicates like that.

I think we should benchmark things. Are we sure that always using the smaller dfs spatial index is still faster?

Reply:
It might be better to put the user in explicit control over the left vs right side of the join based on how they call the function, rather than trying to guess at the correct way to optimize left vs right within the function.

Which one is faster seems very much specific to the geometric relationship between the bounding boxes of the left side vs right side, and if predicates are involved, also where the use of prepared geometries helps speed up evaluation (usually yes for polygons, no speedup for points).  In my own work using `query_bulk` for spatial join, I'll explore doing it both ways and pick the fastest.

I think in general the time is usually a function of the size of the outer loop of all overlaps between the bounding boxes of the left geometries and the right geometries.  Where this loop is smaller in size, we'd expect it to be faster.

In general, IF the size of the left side of the join is smaller than the right, and the geometries on the right do not overlap multiple geometries on the left, then picking the smaller set as the left side will likely be faster.  Example:
- small number of non-overlapping polygons on the left, many points on the right
- small number of non-overlapping polygons on the left, larger number of non-overlapping polygons on the right

In these cases, the left side spatially partitions the right, so it ends up working out faster.

This would likely fail if there were substantial overlaps in the left side geometries, as you'd end up testing the same right side geometries repeatedly against the left side; then it is a question of which join direction produces fewer combinations, and which side benefits more from prepared geometries.

I think where a user is trying to optimize the performance of their `sjoin`, they can bring knowledge of these specific relationships and underlying geometry types to their use of this function.  In that case, perhaps we can handle this instead via guidance in the docs and examples: "if you are trying to spatial join a small number of polygons to a large number of points..."  It is harder to provide this guidance where there is variation in the underlying approach outside of the user's control.

Reply:
Great input @brendan-ward 

I agree that if we put the user in control and directly map the dataframes to the query/predicate operations, that would simplify our work here and might even give users some speedups or further understanding of their data.

Ultimately, it would be nice if `sjoin` just wrapped the spatial index query and predicate checks with logic for the handling the dataframe level join (`how` parameter) as well as converting integer/iloc indexes (used by spatial queries) to dataframe indexes (which might be loc/label).

Reply:
>>This part confuses me a bit. It would be great if someone could confirm if I am right or wrong.

> As far as I understand, this would need to depend on whether the spatial index of the left or right frame is being used

I don't think so. Contains is faster operation than within. And because they are interchangeable (you just change the perspective), sjoin does always contains and never within. Just flips the perspective. That is what happens in the original code above. I don't think your code does the same.

Reply:
@martinfleis is that accurate for both shapely and pygeos, independently of the geometries?

Using benchmarks copied from https://github.com/geopandas/geopandas/pull/1401#issuecomment-622583016:
| rtree      | pygeos      | ratio | test                                              |
|------------|-------------|-------|---------------------------------------------------|
| 27.0±2ms   | 20.9±0.5ms  | ~0.77 | time_query_bulk('contains', 'points', 'polygons') |
| 30.2±2ms   | 150±7μs     | 0.00  | time_query_bulk('contains', 'polygons', 'points') |
| 28.3±0.6ms | 791±20μs    | 0.03  | time_query_bulk('within', 'points', 'polygons')   |
| 28.2±0.5ms | 1.33±0.03ms | 0.05  | time_query_bulk('within', 'polygons', 'points')   |

For rtree, there is not much difference here, maybe the time complexity was dominated by the tree search (or some other operation) instead of the predicate comparisons.

For pygeos, if we are comparing `polygons` to `points` using `op=within` (user requested operation: `points.query_bulk(polygons, 'within')`), swapping for `contains` would actually slow things down ~20x! Is there something I am missing here?